File: app/javascript/channels/delete_channel.js
import consumer from "./consumer"
$(function(){

	const chat = consumer.subscriptions.create({channel: "DeleteChannel", room: $("#messages").data('room_id'), current_user: $('#messages').data('current_user') }, {
	  connected() {
	    // Called when the subscription is ready for use on the server
	  },

	  disconnected() {
	    // Called when the subscription has been terminated by the server
	  },

	  received(data) {
	    // Called when there's incoming data on the websocket for this channel
	    if($('#messages').data('current_user') == data['user_id'] && data['message_reply'] == false){
	    	$("#message-support-" + data['id']).remove();
	    }else if($("#messages").data('current_user') == data['user_id'] && data['message_reply'] == true){
	    	$('#message_reply-support-' + data['id']).remove();
	    }
	  },
	  delete: function(id){
	  	return this.perform('delete', {
	  		id: id
	  	})
	  },
	  delete_reply: function(id){
	  	return this.perform('delete_reply', {
	  		id: id
	  	})
	  }
	});

	$(document).on('click', '.delete_btn', function(event){
		if(window.confirm("本当に削除しますか？")){
			chat.delete(event.target.id);

		}else{
			return false;
		}

	});
	$(document).on('click', '.delete_btn_reply', function(event){
		if(window.confirm("本当に削除しますか？")){
			chat.delete_reply(event.target.dataset.id);
		}else{
			return false;
		}
	});
});

File: app/channels/delete_channel.rb
class DeleteChannel < ApplicationCable::Channel
  def subscribed
    stream_from "delete_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def delete(data)
    if current_user.nil?
      return false
    end
    if Message.find_by(id: data['id'], login: false).present?
  	 return false
    end
    if current_user.present?
  		message = Message.find_by(id: data['id'], room_id: params['room'], user_id: current_user.id, login: true)
  	end
  	# もしmessageに当てはまらなかったら強制的にはじき返す
    # binding.pry
  	if message.nil?
  		return false
  	end
    if message.present?
  	 message.destroy!
    end
    ActionCable.server.broadcast 'delete_channel', id: data['id'], room_id: params['room'], user_id: message.user_id, message_reply: false
  end

  def delete_reply(data)
  	if current_user.nil?
  		return false
  	end
  	if current_user.present?
  		if MessageReply.find_by(id: data['id'], login: false).present?
  			return false
  		end
  		message = MessageReply.find_by(id: data['id'], room_id: params['room'], user_id: current_user.id, login: true)
  		if message.nil?
  			return false
  		elsif message.present?
  			message.destroy!
  		end

  		ActionCable.server.broadcast "delete_channel", id: data['id'], room_id: params['room'], user_id: current_user.id, message_reply: true
  	end
  end
end
